examples/RandomParams.jl

```diff
@@ -0,0 +1,653 @@
+  end
+  width = div(range.stop-range.start, range.step)  # should be exact division
+  if range.start > 0
+    return (range.start):(range.step):(range.stop + range.step*width)
+  end
+  if range.stop < 0
+    return (range.start - range.step*width):(range.step):(range.stop)
+  end
+  half_width = div(1+width,2)
+  delta = range.step*half_width
+  return (range.start-delta):(range.step):(range.stop+delta)
+end
+
+function increase_deg_range(range::StepRange{Int,Int})
+  return (range.start):(range.step):(range.stop+range.step)  # !! should check for overflow !!
+end
+
+
+# Default: always gives false
+function is_compatible(::RandomParams, ::Ring)
+  return false
+end
+
+
+function RAND(R::Ring)
+  # This should always work if the ring R is finite!
+  # Maybe check whether a global flag "PhysicistMode" is set?
+  @req (physicist_mode || is_finite(R))  "Please specify a rand_range, or activate physicist mode"
+  return RAND(rand_range(R))
+end
+
+function rand_range(R::Ring)
+  error("Not (yet) defined")
+end
+
+
+#-------------------------------------------------------
+# Integers
+
+mutable struct RandomParams_ZZ  <: RandomParams
+  range::StepRange{ZZRingElem,ZZRingElem}
+  nz::Bool
+  counter::Int64  # incremented each time a new random value is generated
+  num_different_values::Int64  #
+
+  function RandomParams_ZZ(lwb_upb::StepRange{ZZRingElem,ZZRingElem}, non_zero::Bool)
+    @req (lwb_upb.start <= lwb_upb.stop)   "Empty range"
+    @req !(non_zero && lwb_upb == 0:0)  "Cannot request non-zero elements from zero range"
+    result = new()
+    result.range = lwb_upb
+    result.nz = non_zero
+    result.counter = 0
+    result.num_different_values = cap_to_Int64(length(result.range))
+    return result
+  end
+end
+
+function is_compatible(::RandomParams_ZZ, ::ZZRing)
+  return true
+end
+
+function rand_range(::ZZRing; ZZ_range::AbstractRange{T}=-99:99, non_zero::Bool=false)  where {T <: Nemo.IntegerUnion}
+  return RandomParams_ZZ(range_ZZ(ZZ_range), non_zero)
+end
+
+# function symmetric_range(::ZZRing, B::ZZRingElem)  # allow more integer types?
+#   @req (B >= 0)  "Empty symmetric range"
+#   return RandomParams_ZZ(-B:ZZ(1):B, false)
+# end
+
+# function nonneg_range(::ZZRing, B::ZZRingElem)  # allow more integer types?
+#   @req (B >= 0)  "Empty non-neg range"
+#   return RandomParams_ZZ(ZZ(0):ZZ(1):B, false)
+# end
+
+
+
+function extend_range!(params::RandomParams_ZZ)
+  params.range = double_range(params.range)
+  ## update counter & num_different_values...
+  params.counter = 0 # ? just reset counter ?
+  params.num_different_values = cap_to_Int64(length(params.range))
+  return params
+end
+
+
+
+function RAND(params::RandomParams_ZZ)
+  params.counter += 1
+  val = ZZ(0)  # just for scoping
+  # The constructor has checked that this loop cannot be infinite
+  while true
+    val = ZZ(rand(params.range))
+    if (!params.nz || !is_zero(val))
+      break;
+    end
+  end
+  return val
```

Comment:
No need to allocate a temp object just for scoping...
```suggestion
  # The constructor has checked that this loop cannot be infinite
  while true
    val = ZZ(rand(params.range))
    if (!params.nz || !is_zero(val))
      return val
    end
  end
```

Reply:
Thanks!  Not sure why my original implementation was so "clunky" -- weird side-effect of "code evolution"?